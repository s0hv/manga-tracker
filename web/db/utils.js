import {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
  UNIQUE_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  FOREIGN_KEY_VIOLATION,
  NOT_NULL_VIOLATION,
} from 'pg-error-constants';

import { pgp } from '.';
import { NoColumnsError } from './errors.js';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';

/**
 * Generate update statement from an object while filtering out undefined values.
 * The generated update statement lacks a WHERE clause and has the values embedded in the query.
 * @param {Object} o Input object
 * @param {String} tableName Name of the table to be updated
 * @returns {String} The update statement without a whereclause
 */
export const generateUpdate = (o, tableName) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);
  try {
    return pgp.helpers.update(obj, null, tableName);
  } catch (e) {
    throw new NoColumnsError('No valid columns given');
  }
};

export function handleError(err, res, msgOverrides = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter(e => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
