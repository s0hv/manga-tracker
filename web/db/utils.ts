import {
  FOREIGN_KEY_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  INVALID_TEXT_REPRESENTATION,
  NOT_NULL_VIOLATION,
  NUMERIC_VALUE_OUT_OF_RANGE,
  UNIQUE_VIOLATION,
} from 'pg-error-constants';
import type { Response } from 'express-serve-static-core';

import { NoColumnsError } from './errors';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';
import type { Db } from '.';

/**
 * Generate update statement from an object while filtering out undefined values.
 * Do not pass untrusted properties to this method, as it will update every column given to it
 * @param {Object} o Input object
 * @param {Db} sql Database instance
 */
export const generateUpdate = (o: any, sql: Db) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);

  if (Object.keys(obj).length === 0) throw new NoColumnsError('No valid columns given');

  return sql(obj);
};

export function handleError(err: any, res: Response, msgOverrides: any = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter((e: any) => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
