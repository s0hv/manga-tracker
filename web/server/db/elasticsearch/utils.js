export const handleElasticError = (err, res) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export const extractFields = (result, fields, idPrefix, customFieldFormatter) => {
  idPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o = {
      [`${idPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field] = (hit.fields[field] || [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => { o[k] = formatted[k] });
      }
    }

    return o;
  });
};
