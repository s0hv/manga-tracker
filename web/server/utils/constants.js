export const csrfMissing = 'Missing CSRF token';
export const isDev = process.env.NODE_ENV !== 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const NO_GROUP = 1;
