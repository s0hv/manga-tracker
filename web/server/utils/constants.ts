export const csrfMissing = 'CSRF error. Modifying requests must come from the same origin.';
export const isDev = process.env.NODE_ENV !== 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const NO_GROUP = 1;

export const isSecure = (process.env.NEXTAUTH_URL || '').startsWith('https://');
export const cookiePrefix = (process.env.NEXTAUTH_URL || '').startsWith('https://') ? '__Secure-' : '';
