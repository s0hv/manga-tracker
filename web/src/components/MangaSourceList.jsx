import React from 'react';
import ExpandMore from '@material-ui/icons/ExpandMore';
import ExpandLess from '@material-ui/icons/ExpandLess';
import {
  Button,
  Collapse,
  Link,
  List,
  ListItem,
  ListItemText,
  Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import { useUser } from '../utils/useUser';


const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    maxWidth: 360,
    backgroundColor: theme.palette.background.paper,
  },
  mainItem: {
    borderRadius: theme.shape.borderRadius,
    backgroundColor: theme.palette.primary.main,
    '&:hover': {
      backgroundColor: theme.palette.primary.dark,
    },
  },
  nested: {
    paddingLeft: theme.spacing(4),
    display: 'flex',
    justifyContent: 'space-between',
  },
  listOpener: {
    color: theme.palette.primary.contrastText,
  },
}));

function MangaSourceList(props) {
  const {
    items = [],
    followUnfollow = (_) => undefined,
    userFollows = [],
    classesProp = [],
    openByDefault = false,
  } = props;

  const { isAuthenticated } = useUser();
  const classes = useStyles();
  const [open, setOpen] = React.useState(openByDefault);
  const handleClick = () => {
    setOpen(!open);
  };

  function renderItem(item) {
    return (
      <ListItem className={classes.nested} key={item.serviceId}>
        <Typography>
          <Link href={item.url.replace('{}', item.titleId)} target='_blank' rel='noopener noreferrer'>{item.name}</Link>
        </Typography>
        {isAuthenticated && (
          <Button
            variant='contained'
            color='primary'
            onClick={followUnfollow(item.serviceId)}
            aria-label={`Follow ${item.name}`}
          >
            {userFollows.indexOf(item.serviceId) < 0 ? 'Follow' : 'Unfollow'}
          </Button>
        )}
      </ListItem>
    );
  }

  return (
    <List
      aria-label='manga sources'
      className={`${classes.root} ${classesProp.join(' ')}`}
    >
      <ListItem
        button
        onClick={handleClick}
        className={classes.mainItem}
        aria-label={`${open ? 'close' : 'open'} follows`}
      >
        <ListItemText primary='Manga sources' className={classes.listOpener} />
        {open ? <ExpandLess /> : <ExpandMore />}
      </ListItem>
      <Collapse in={open} timeout='auto' unmountOnExit>
        <List component='div' disablePadding aria-hidden={!open}>
          {items.map((item, index) => renderItem(item, index))}
        </List>
      </Collapse>
    </List>
  );
}

MangaSourceList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object),
  followUnfollow: PropTypes.func,
  userFollows: PropTypes.arrayOf(PropTypes.number),
  classesProp: PropTypes.arrayOf(PropTypes.string),
  openByDefault: PropTypes.bool,
};

export default MangaSourceList;
