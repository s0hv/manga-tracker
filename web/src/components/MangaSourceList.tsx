import React, { MouseEventHandler } from 'react';
import ExpandMore from '@mui/icons-material/ExpandMore';
import ExpandLess from '@mui/icons-material/ExpandLess';
import {
  Button,
  Collapse,
  Link,
  List,
  ListItem,
  ListItemButton,
  ListItemText,
  Typography,
} from '@mui/material';
import { useUser } from '../utils/useUser';
import type { MangaServiceData } from '@/types/api/manga';


export type MangaSourceListProps = {
  items?: MangaServiceData[]
  followUnfollow?: (serviceId: number | null) => (undefined | MouseEventHandler<HTMLButtonElement>)
  userFollows?: (number | null)[]
  classesProp?: string[]
  openByDefault?: boolean
}
function MangaSourceList(props: MangaSourceListProps) {
  const {
    items = [],
    followUnfollow = (_) => undefined,
    userFollows = [],
    classesProp = [],
    openByDefault = false,
  } = props;

  const { isAuthenticated } = useUser();
  const [open, setOpen] = React.useState(openByDefault);
  const handleClick = () => {
    setOpen(!open);
  };

  function renderItem(item: MangaServiceData) {
    const followText = userFollows.indexOf(item.serviceId) < 0 ? 'Follow' : 'Unfollow';
    return (
      <ListItem key={item.serviceId} sx={{ pl: 4, display: 'flex', justifyContent: 'space-between' }}>
        <Typography>
          <Link
            href={item.url.replace('{}', item.titleId)}
            target='_blank'
            rel='noopener noreferrer'
            underline='hover'
          >
            {item.name}
          </Link>
        </Typography>
        {isAuthenticated && (
          <Button
            variant='contained'
            color='primary'
            onClick={followUnfollow(item.serviceId)}
            aria-label={`${followText.toLowerCase()} ${item.name}`}
          >
            {followText}
          </Button>
        )}
      </ListItem>
    );
  }

  return (
    <List
      aria-label='manga sources'
      sx={{ width: '100%', maxWidth: '360px' }}
      className={`${classesProp.join(' ')}`}
    >
      <ListItemButton
        onClick={handleClick}
        aria-label={`${open ? 'close' : 'open'} follows`}
        sx={{
          borderRadius: 1,
          backgroundColor: 'primary.main',
          '&:hover': {
            backgroundColor: 'primary.dark',
          },
        }}
      >
        <ListItemText primary='Manga sources' sx={{ color: 'primary.contrastText' }} />
        {open ? <ExpandLess /> : <ExpandMore />}
      </ListItemButton>
      <Collapse in={open} timeout='auto' unmountOnExit>
        <List component='div' disablePadding aria-hidden={!open}>
          {items.map((item) => renderItem(item))}
        </List>
      </Collapse>
    </List>
  );
}

export default MangaSourceList;
