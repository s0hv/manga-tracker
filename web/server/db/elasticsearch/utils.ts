import { ApiResponse } from '@elastic/elasticsearch';
import type { Response } from 'express-serve-static-core';

import type {
  MangaSearchResult,
  MangaSearchResultFields,
} from '@/db/elasticsearch/manga';

export const handleElasticError = (err: unknown, res: Response) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export type ExtractedFields = {
  score: number
  [key: string]: unknown
};
export type CustomFieldFormatter<T extends boolean> = (fields: MangaSearchResultFields<T>) => Record<string, unknown>;

export const extractFields = <T extends boolean>(
  result: ApiResponse<MangaSearchResult<T>>,
  fields?: (keyof MangaSearchResultFields<T>)[],
  idPrefix?: string,
  customFieldFormatter?: (fields: MangaSearchResultFields<T>) => Record<string, unknown>
) => {
  const mappedIdPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o: ExtractedFields = {
      [`${mappedIdPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field as string] = (hit.fields[field] as Array<unknown> ?? [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => {
          o[k] = formatted[k];
        });
      }
    }

    return o;
  });
};
