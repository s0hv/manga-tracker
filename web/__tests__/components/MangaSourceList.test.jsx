import React from 'react';
import { render, screen } from '@testing-library/react';

import 'jest-extended';

import userEvent from '@testing-library/user-event';
import MangaSourceList from '../../src/components/MangaSourceList';
import { normalUser, withUser } from '../utils';

const services = [
  {
    titleId: '100010',
    serviceId: 1,
    name: 'MANGA Plus',
    url: 'https://mangaplus.shueisha.co.jp/titles/{}',
  },
  {
    titleId: '20882',
    serviceId: 2,
    name: 'MangaDex',
    url: 'https://mangadex.org/title/{}',
  },
  {
    titleId: 'test_series_1',
    serviceId: 3,
    name: 'TestService',
    url: 'https://test.com/manga/{}',
  },
];

const follows = [1];


describe('Manga source list', () => {
  it('Should render correctly without input', () => {
    render(<MangaSourceList />);

    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
  });

  it('Should render correctly with items without user', () => {
    render(
      <MangaSourceList
        items={services}
        classesProp={['test-class-1', 'test-class-2']}
        openByDefault
      />
    );

    expect(screen.queryAllByRole('listitem')).toHaveLength(services.length);
    expect(screen.queryByRole('button', { name: /follow /i })).not.toBeInTheDocument();
  });

  it('Should render correctly with items and user', async () => {
    render(
      await withUser(
        normalUser, (
          <MangaSourceList
            items={services}
            userFollows={follows}
            classesProp={['test-class-1', 'test-class-2']}
            openByDefault
          />
        )
      )
    );

    expect(screen.queryAllByRole('listitem')).toHaveLength(services.length);
    expect(screen.queryAllByRole('button', { name: /follow /i })).toHaveLength(services.length);
  });

  it('should fail to render with invalid services', () => {
    // At the moment missing name or service id don't throw errors
    const invalid = [{
      titleId: 'test_series_1',
      serviceId: 3,
    }];

    expect(() => render(
      <MangaSourceList
        items={invalid}
        classesProp={['test-class-1', 'test-class-2']}
      />
    )).toThrow();
  });
});


describe('Manga source list should handle user input', () => {
  it('Should call follow function on click', async () => {
    const followUnfollow = jest.fn();
    const createEvent = jest.fn();
    createEvent.mockImplementation(() => followUnfollow);

    render(
      await withUser(
        normalUser, (
          <MangaSourceList
            items={services}
            userFollows={follows}
            classesProp={['test-class-1', 'test-class-2']}
            followUnfollow={createEvent}
            openByDefault
          />
        )
      )
    );

    expect(createEvent).toHaveBeenCalledTimes(services.length);
    expect(followUnfollow).toHaveBeenCalledTimes(0);

    userEvent.click(
      screen.getByRole(
        'button',
        { name: new RegExp(`follow ${services[0].name}`, 'i') }
      )
    );
    expect(followUnfollow).toHaveBeenCalledTimes(1);
  });

  it('Should open and close collapsed list on clicks', async () => {
    render(
      await withUser(
        normalUser, (
          <MangaSourceList
            items={services}
            userFollows={follows}
            classesProp={['test-class-1', 'test-class-2']}
          />
        )
      )
    );

    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();

    userEvent.click(
      screen.getByRole('button', { name: /open follows/i })
    );

    expect(screen.queryAllByRole('button', { name: /follow /i })).toHaveLength(services.length);

    userEvent.click(
      screen.getByRole('button', { name: /close follows/i })
    );

    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
  });
});
