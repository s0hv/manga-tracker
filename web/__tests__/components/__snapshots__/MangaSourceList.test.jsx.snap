// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manga source list Should render correctly with items and user 1`] = `
<nav
  aria-label="manga sources"
  className="MuiList-root makeStyles-root-1 test-class-1 test-class-2 MuiList-padding"
>
  <div
    aria-disabled={false}
    className="MuiButtonBase-root MuiListItem-root makeStyles-mainItem-2 MuiListItem-gutters MuiListItem-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemText-root makeStyles-listOpener-4"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Manga sources
      </span>
    </div>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root"
    />
  </div>
  <div
    className="MuiCollapse-container MuiCollapse-entered"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      >
        <div
          className="MuiList-root"
        >
          <li
            className="MuiListItem-root makeStyles-nested-3 MuiListItem-gutters"
            disabled={false}
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://mangaplus.shueisha.co.jp/titles/100010"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noopener noreferrer"
                target="_blank"
              >
                MANGA Plus
              </a>
            </p>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              disabled={false}
              name="Follow MANGA Plus"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Unfollow
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li
            className="MuiListItem-root makeStyles-nested-3 MuiListItem-gutters"
            disabled={false}
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://mangadex.org/title/20882"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noopener noreferrer"
                target="_blank"
              >
                MangaDex
              </a>
            </p>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              disabled={false}
              name="Follow MangaDex"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Follow
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li
            className="MuiListItem-root makeStyles-nested-3 MuiListItem-gutters"
            disabled={false}
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://test.com/manga/test_series_1"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noopener noreferrer"
                target="_blank"
              >
                TestService
              </a>
            </p>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              disabled={false}
              name="Follow TestService"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Follow
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </li>
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`Manga source list Should render correctly with items without user 1`] = `
<nav
  aria-label="manga sources"
  className="MuiList-root makeStyles-root-1 test-class-1 test-class-2 MuiList-padding"
>
  <div
    aria-disabled={false}
    className="MuiButtonBase-root MuiListItem-root makeStyles-mainItem-2 MuiListItem-gutters MuiListItem-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemText-root makeStyles-listOpener-4"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Manga sources
      </span>
    </div>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root"
    />
  </div>
  <div
    className="MuiCollapse-container MuiCollapse-entered"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      >
        <div
          className="MuiList-root"
        >
          <li
            className="MuiListItem-root makeStyles-nested-3 MuiListItem-gutters"
            disabled={false}
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://mangaplus.shueisha.co.jp/titles/100010"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noopener noreferrer"
                target="_blank"
              >
                MANGA Plus
              </a>
            </p>
          </li>
          <li
            className="MuiListItem-root makeStyles-nested-3 MuiListItem-gutters"
            disabled={false}
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://mangadex.org/title/20882"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noopener noreferrer"
                target="_blank"
              >
                MangaDex
              </a>
            </p>
          </li>
          <li
            className="MuiListItem-root makeStyles-nested-3 MuiListItem-gutters"
            disabled={false}
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://test.com/manga/test_series_1"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noopener noreferrer"
                target="_blank"
              >
                TestService
              </a>
            </p>
          </li>
        </div>
      </div>
    </div>
  </div>
</nav>
`;
